tpu/bootstrap.sh

```diff
@@ -0,0 +1,87 @@
+#! bin/bash
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Configures a single-client v2-8 TPU VM for DTensor.
+
+# This command produces cluster-run.sh, which can be used to run command
+# on the TPU VM.
+#
+# The git repo is cloned to the VMs.
+
+NAME="${USER}-dtensor-tpu-test"
+ZONE=europe-west4-a
+VERSION="tpu-vm-tf-2.9.1"
+TOPOLOGY=$1
+NUM_CORES=$(cut -d "-" -f2- <<< $TOPOLOGY)
+export NUM_WORKERS=$(($NUM_CORES / 8))
+export GCS_BUCKET=${GCS_BUCKET:-dtensor-checkpoints}
+
+if [ $NUM_WORKERS -gt 1 ]; then
+  VERSION+="-pod"
+fi
+
+bash `dirname $0`/make-cluster-commands.sh "${NAME}" "${ZONE}" "${NUM_WORKERS}"
+
+set -x
+gcloud compute tpus tpu-vm create $NAME  \
+    --zone=$ZONE  \
+    --accelerator-type=$TOPOLOGY  \
+    --version=$VERSION
+
+gcloud compute firewall-rules create allow-ssh --allow=tcp:22
+set +x
+
+while bash cluster-run.sh ls |grep 'exited with return code'; do
+  echo Health checking
+  sleep 10
+done
+
+bash cluster-run.sh "if ! [[ -d dtensor-gcp-examples ]]; then git clone https://github.com/tensorflow/dtensor-gcp-examples; fi"
+bash cluster-run.sh "cd dtensor-gcp-examples; git pull;"
+bash cluster-run.sh "ls -l dtensor-gcp-examples;"
+
+cat > launch <<EOF
+#! /bin/bash
+
+export TF_CPP_MIN_LOG_LEVEL=3
+export DTENSOR_CLIENT_ID="\${HOSTNAME: -1}"
+export DTENSOR_NUM_CLIENTS=$NUM_WORKERS
+export DTENSOR_JOB_NAME=worker
+WORKER_NAME="\${HOSTNAME:: -2}"
+PORT=19011
```

Comment:
Seems to be a magic number, and the port might not be available when landed on the worker.

Reply:
Is there a solution for this to pick an available port number? I looked at the GPU launch script and even in there we are picking a random port number.

Reply:
In this case, you might need to run some test to ensure the port is available on each host before using that. If you would like to take the approach of each worker to report the port it will use, then the script need further work since it need to concat the host:port together and pass to each worker.

Reply:
Adding to Yu's response, since this repository only holds example scripts I feel adding logic here to handle already-in-use ports might distract from the other setup details we want to communicate. Port selection is something we will likely handle in the coordination service.